<a href="https://colab.research.google.com/github/Rabiatou08/DI-Bootcamp/blob/main/week2/Day1/ExerciseXP/exercicesxp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Semaine 2 - Jour 1 : Exercices XP

In [ ]:
from google.colab import files
uploaded = files.upload()

## Exercice 1 : Introduction à l'analyse des données

### Qu'est-ce que l'analyse de données ?
L'analyse de données est le processus de collecte, nettoyage
et interprétation de données brutes pour la prise de décision.

### Pourquoi est-elle importante ?
Elle permet de prendre des décisions éclairées, réduire les
risques et optimiser les processus.

### Trois domaines d'application
1. Santé : prédire la propagation des maladies
2. Finance : détecter les fraudes bancaires
3. Marketing : personnaliser les campagnes publicitaires

## Exercice 2 : Chargement et analyse initiale

In [2]:
import pandas as pd
import zipfile

# Dataset 1 : Sommeil
df1 = pd.read_csv('How Much Sleep Do Americans.zip', compression='zip')
display(df1.head())
print(df1.shape)
df1.info()
display(df1.describe())

,index,Year,Period,Avg hrs per day sleeping,Standard Error,Type of Days,Age Group,Activity,Sex
0,0,2003,Annual,8.57,0.018,All days,15 years and over,Sleeping,Both
1,1,2004,Annual,8.55,0.026,All days,15 years and over,Sleeping,Both
2,2,2005,Annual,8.62,0.023,All days,15 years and over,Sleeping,Both
3,3,2006,Annual,8.63,0.024,All days,15 years and over,Sleeping,Both
4,4,2007,Annual,8.57,0.024,All days,15 years and over,Sleeping,Both


(945, 9)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 945 entries, 0 to 944
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     945 non-null    int64  
 1   Year                      945 non-null    int64  
 2   Period                    945 non-null    object 
 3   Avg hrs per day sleeping  945 non-null    float64
 4   Standard Error            945 non-null    float64
 5   Type of Days              945 non-null    object 
 6   Age Group                 945 non-null    object 
 7   Activity                  945 non-null    object 
 8   Sex                       945 non-null    object 
dtypes: float64(2), int64(2), object(5)
memory usage: 66.6+ KB


,index,Year,Avg hrs per day sleeping,Standard Error
count,945.000000,945.000000,945.000000,945.000000
mean,472.000000,2010.000000,8.806868,0.086958
std,272.942302,4.322782,0.545001,0.034631
min,0.000000,2003.000000,7.720000,0.018000
25%,236.000000,2006.000000,8.360000,0.063000
50%,472.000000,2010.000000,8.810000,0.084000
75%,708.000000,2014.000000,9.190000,0.106000
max,944.000000,2017.000000,10.650000,0.228000


### Description Dataset 1 : Sommeil des Américains
- 945 lignes, 9 colonnes
- Données de 2003 à 2017
- Habitudes de sommeil par âge, sexe et année
- Moyenne de sommeil : 8.8h par jour

In [3]:
# Dataset 2 : Santé mentale
df2 = pd.read_csv('Global Trends in Mental Health Disorder.zip', compression='zip')
display(df2.head())
print(df2.shape)
df2.info()
display(df2.describe())

/tmp/ipykernel_2645/2969716966.py:2: DtypeWarning: Columns (5,6) have mixed types. Specify dtype option on import or set low_memory=False.
  df2 = pd.read_csv('Global Trends in Mental Health Disorder.zip', compression='zip')


,index,Entity,Code,Year,Schizophrenia (%),Bipolar disorder (%),Eating disorders (%),Anxiety disorders (%),Drug use disorders (%),Depression (%),Alcohol use disorders (%)
0,0,Afghanistan,AFG,1990,0.16056,0.697779,0.101855,4.828830,1.677082,4.071831,0.672404
1,1,Afghanistan,AFG,1991,0.160312,0.697961,0.099313,4.829740,1.684746,4.079531,0.671768
2,2,Afghanistan,AFG,1992,0.160135,0.698107,0.096692,4.831108,1.694334,4.088358,0.670644
3,3,Afghanistan,AFG,1993,0.160037,0.698257,0.094336,4.830864,1.705320,4.096190,0.669738
4,4,Afghanistan,AFG,1994,0.160022,0.698469,0.092439,4.829423,1.716069,4.099582,0.669260


(108553, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108553 entries, 0 to 108552
Data columns (total 11 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   index                      108553 non-null  int64  
 1   Entity                     108553 non-null  object 
 2   Code                       103141 non-null  object 
 3   Year                       108553 non-null  object 
 4   Schizophrenia (%)          25875 non-null   object 
 5   Bipolar disorder (%)       19406 non-null   object 
 6   Eating disorders (%)       100236 non-null  object 
 7   Anxiety disorders (%)      6468 non-null    float64
 8   Drug use disorders (%)     6468 non-null    float64
 9   Depression (%)             6468 non-null    float64
 10  Alcohol use disorders (%)  6468 non-null    float64
dtypes: float64(4), int64(1), object(6)
memory usage: 9.1+ MB


,index,Anxiety disorders (%),Drug use disorders (%),Depression (%),Alcohol use disorders (%)
count,108553.000000,6468.000000,6468.000000,6468.000000,6468.000000
mean,54276.000000,3.989921,0.862278,3.497654,1.585821
std,31336.696223,1.167526,0.460679,0.655859,0.860283
min,0.000000,2.023393,0.383650,2.139903,0.446940
25%,27138.000000,3.188824,0.535064,3.005529,0.993685
50%,54276.000000,3.554373,0.726430,3.499606,1.479936
75%,81414.000000,4.682163,0.940157,3.912381,1.867834
max,108552.000000,8.967330,3.452476,6.602754,5.474668


### Description Dataset 2 : Santé mentale mondiale
- 108 553 lignes, 11 colonnes
- Données par pays de 1990 à aujourd'hui
- Contient les % de troubles mentaux : dépression, anxiété...

In [4]:
# Dataset 3 : Cartes de crédit
with zipfile.ZipFile('Credit Card Approvals (Clean Data).zip') as z:
    df3 = pd.read_csv(z.open('Credit Card Approvals (Clean Data)/clean_dataset.csv'))
display(df3.head())
print(df3.shape)
df3.info()
display(df3.describe())

,Gender,Age,Debt,Married,BankCustomer,Industry,Ethnicity,YearsEmployed,PriorDefault,Employed,CreditScore,DriversLicense,Citizen,ZipCode,Income,Approved
0,1,30.83,0.000,1,1,Industrials,White,1.25,1,1,1,0,ByBirth,202,0,1
1,0,58.67,4.460,1,1,Materials,Black,3.04,1,1,6,0,ByBirth,43,560,1
2,0,24.50,0.500,1,1,Materials,Black,1.50,1,0,0,0,ByBirth,280,824,1
3,1,27.83,1.540,1,1,Industrials,White,3.75,1,1,5,1,ByBirth,100,3,1
4,1,20.17,5.625,1,1,Industrials,White,1.71,1,0,0,0,ByOtherMeans,120,0,1


(690, 16)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          690 non-null    int64  
 1   Age             690 non-null    float64
 2   Debt            690 non-null    float64
 3   Married         690 non-null    int64  
 4   BankCustomer    690 non-null    int64  
 5   Industry        690 non-null    object 
 6   Ethnicity       690 non-null    object 
 7   YearsEmployed   690 non-null    float64
 8   PriorDefault    690 non-null    int64  
 9   Employed        690 non-null    int64  
 10  CreditScore     690 non-null    int64  
 11  DriversLicense  690 non-null    int64  
 12  Citizen         690 non-null    object 
 13  ZipCode         690 non-null    int64  
 14  Income          690 non-null    int64  
 15  Approved        690 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 86.4+ KB


,Gender,Age,Debt,Married,BankCustomer,YearsEmployed,PriorDefault,Employed,CreditScore,DriversLicense,ZipCode,Income,Approved
count,690.000000,690.000000,690.000000,690.000000,690.000000,690.000000,690.000000,690.000000,690.00000,690.000000,690.000000,690.000000,690.000000
mean,0.695652,31.514116,4.758725,0.760870,0.763768,2.223406,0.523188,0.427536,2.40000,0.457971,180.547826,1017.385507,0.444928
std,0.460464,11.860245,4.978163,0.426862,0.425074,3.346513,0.499824,0.495080,4.86294,0.498592,173.970323,5210.102598,0.497318
min,0.000000,13.750000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,22.670000,1.000000,1.000000,1.000000,0.165000,0.000000,0.000000,0.00000,0.000000,60.000000,0.000000,0.000000
50%,1.000000,28.460000,2.750000,1.000000,1.000000,1.000000,1.000000,0.000000,0.00000,0.000000,160.000000,5.000000,0.000000
75%,1.000000,37.707500,7.207500,1.000000,1.000000,2.625000,1.000000,1.000000,3.00000,1.000000,272.000000,395.500000,1.000000
max,1.000000,80.250000,28.000000,1.000000,1.000000,28.500000,1.000000,1.000000,67.00000,1.000000,2000.000000,100000.000000,1.000000


### Description Dataset 3 : Cartes de crédit
- Données sur les approbations/refus de cartes de crédit
- Contient des infos financières et démographiques

## Exercice 3 : Identification des types de données

In [5]:
print("=== Dataset 1 : Sommeil ===")
print(df1.dtypes)
print("\n=== Dataset 2 : Santé mentale ===")
print(df2.dtypes)
print("\n=== Dataset 3 : Cartes de crédit ===")
print(df3.dtypes)

=== Dataset 1 : Sommeil ===
index                         int64
Year                          int64
Period                       object
Avg hrs per day sleeping    float64
Standard Error              float64
Type of Days                 object
Age Group                    object
Activity                     object
Sex                          object
dtype: object

=== Dataset 2 : Santé mentale ===
index                          int64
Entity                        object
Code                          object
Year                          object
Schizophrenia (%)             object
Bipolar disorder (%)          object
Eating disorders (%)          object
Anxiety disorders (%)        float64
Drug use disorders (%)       float64
Depression (%)               float64
Alcohol use disorders (%)    float64
dtype: object

=== Dataset 3 : Cartes de crédit ===
Gender              int64
Age               float64
Debt              float64
Married             int64
BankCustomer        int64
Industry  

### Dataset 1 : Sommeil
| Colonne                  | Type         | Raison            |
|--------------------------|--------------|-------------------|
| Year                     | Quantitative | int64 - nombre    |
| Avg hrs per day sleeping | Quantitative | float64 - décimal |
| Standard Error           | Quantitative | float64 - décimal |
| Period                   | Qualitative  | object - texte    |
| Type of Days             | Qualitative  | object - catégorie|
| Age Group                | Qualitative  | object - catégorie|
| Activity                 | Qualitative  | object - texte    |
| Sex                      | Qualitative  | object - catégorie|

### Dataset 2 : Santé mentale
| Colonne                   | Type         | Raison            |
|---------------------------|--------------|-------------------|
| Year                      | Quantitative | int64 - nombre    |
| Schizophrenia (%)         | Quantitative | float64 - décimal |
| Bipolar disorder (%)      | Quantitative | float64 - décimal |
| Eating disorders (%)      | Quantitative | float64 - décimal |
| Anxiety disorders (%)     | Quantitative | float64 - décimal |
| Drug use disorders (%)    | Quantitative | float64 - décimal |
| Depression (%)            | Quantitative | float64 - décimal |
| Alcohol use disorders (%) | Quantitative | float64 - décimal |
| Entity                    | Qualitative  | object - pays     |
| Code                      | Qualitative  | object - code pays|

### Dataset 3 : Cartes de crédit
→ Catégoriser après exécution de df3.dtypes :
- int64 / float64 → Quantitative
- object → Qualitative

## Exercice 4 : Exploration des types de données - Iris

In [6]:
df4 = pd.read_csv('Iris Dataset.zip', compression='zip')
display(df4.head())
print("Shape :", df4.shape)
print("\nTypes de colonnes :")
print(df4.dtypes)
qualitative = df4.select_dtypes(include=['object']).columns
print("\nColonnes qualitatives :", list(qualitative))
quantitative = df4.select_dtypes(include=['float64', 'int64']).columns
print("Colonnes quantitatives :", list(quantitative))

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


Shape : (150, 6)

Types de colonnes :
Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object

Colonnes qualitatives : ['Species']
Colonnes quantitatives : ['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm']


### Réponse Iris
| Colonne      | Type         | Raison                 |
|--------------|--------------|------------------------|
| sepal_length | Quantitative | float64 - mesure en cm |
| sepal_width  | Quantitative | float64 - mesure en cm |
| petal_length | Quantitative | float64 - mesure en cm |
| petal_width  | Quantitative | float64 - mesure en cm |
| species      | Qualitative  | object - 3 catégories  |

## Exercice 5 : Compétences de base en observation

In [7]:
df5 = pd.read_csv('How Much Sleep Do Americans.zip', compression='zip')
display(df5.head())
print(df5.columns)

,index,Year,Period,Avg hrs per day sleeping,Standard Error,Type of Days,Age Group,Activity,Sex
0,0,2003,Annual,8.57,0.018,All days,15 years and over,Sleeping,Both
1,1,2004,Annual,8.55,0.026,All days,15 years and over,Sleeping,Both
2,2,2005,Annual,8.62,0.023,All days,15 years and over,Sleeping,Both
3,3,2006,Annual,8.63,0.024,All days,15 years and over,Sleeping,Both
4,4,2007,Annual,8.57,0.024,All days,15 years and over,Sleeping,Both


Index(['index', 'Year', 'Period', 'Avg hrs per day sleeping', 'Standard Error',
       'Type of Days', 'Age Group', 'Activity', 'Sex'],
      dtype='object')


### Colonnes pertinentes pour l'analyse

**Analyse de tendances :**
- Year → voir l'évolution du sommeil au fil des années

**Comparaison de groupes :**
- Age Group → comparer le sommeil selon l'âge
- Sex → comparer hommes vs femmes
- Type of Days → comparer jours de semaine vs weekend

**Mesure principale :**
- Avg hrs per day sleeping → la valeur à analyser

## Exercice 6 : Identification des types de données

| Source de données                                     | Type          | Raison                           |
|-------------------------------------------------------|---------------|----------------------------------|
| Rapports financiers stockés dans Excel                | Structuré     | Lignes et colonnes organisées    |
| Photographies sur réseaux sociaux                     | Non structuré | Fichiers images sans format fixe |
| Collection d'articles de presse                       | Non structuré | Texte libre sans structure       |
| Données d'inventaire en base de données relationnelle | Structuré     | Tables organisées en SQL         |
| Enregistrements d'entretiens                          | Non structuré | Audio/texte libre                |

## Exercice 7 : Transformation des données non structurées

| Source               | Méthode de conversion                              | Justification                          |
|----------------------|----------------------------------------------------|----------------------------------------|
| Articles de blog     | NLP → extraire titre, date, mots-clés → CSV        | Le texte doit être analysé avant stockage |
| Enregistrements audio| Speech-to-Text → extraire durée, sujet → CSV       | L'audio doit être converti en texte    |
| Notes manuscrites    | OCR → extraire les idées → Excel                   | Scanner + OCR pour lire le manuscrit   |
| Tutoriel vidéo       | Analyse vidéo → extraire titre, étapes → CSV       | La vidéo doit être analysée avant stockage |

## Exercice 8 : Importer un fichier depuis GitHub

In [8]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df8 = pd.read_csv(url)
display(df8.head())
print(df8.shape)
df8.info()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


(891, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Description Dataset Titanic
- 891 lignes, 12 colonnes
- Données sur les passagers du Titanic
- Colonnes : PassengerId, Survived, Pclass, Name, Sex, Age...

## Exercice 9 : Exporter un dataframe Excel et JSON

In [9]:
!pip install openpyxl
import pandas as pd

df9 = pd.DataFrame({
    'Nom': ['Rabi', 'Ridi', 'Grace'],
    'Age': [22, 3, 20],
    'Ville': ['Abidjan', 'Yakro', 'Bouaké']
})
display(df9)

df9.to_excel('export.xlsx', index=False)
print(" Exporté en Excel !")

df9.to_json('export.json')
print(" Exporté en JSON !")

,Nom,Age,Ville
0,Rabi,22,Abidjan
1,Ridi,3,Yakro
2,Grace,20,Bouaké


✅ Exporté en Excel !
✅ Exporté en JSON !


## Exercice 10 : Lecture de données JSON

In [10]:
import pandas as pd

url = "https://jsonplaceholder.typicode.com/posts"
df10 = pd.read_json(url)
display(df10.head(5))
print(df10.shape)
df10.info()

,userId,id,title,body
0,1,1,sunt aut facere repellat provident occaecati e...,quia et suscipit\nsuscipit recusandae consequu...
1,1,2,qui est esse,est rerum tempore vitae\nsequi sint nihil repr...
2,1,3,ea molestias quasi exercitationem repellat qui...,et iusto sed quo iure\nvoluptatem occaecati om...
3,1,4,eum et est occaecati,ullam et saepe reiciendis voluptatem adipisci\...
4,1,5,nesciunt quas odio,repudiandae veniam quaerat sunt sed\nalias aut...


(100, 4)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   userId  100 non-null    int64 
 1   id      100 non-null    int64 
 2   title   100 non-null    object
 3   body    100 non-null    object
dtypes: int64(2), object(2)
memory usage: 3.3+ KB


### Description Dataset JSON
- 100 lignes
- Colonnes : userId, id, title, body
- Données de posts fictifs depuis une API